src/main/src/main/java/org/geoserver/catalog/CoverageViewReader.java

```diff
@@ -507,6 +411,196 @@ public GridCoverage2D read(GeneralParameterValue[] parameters)
+            CoveragesConsistencyChecker checker)
+            throws IOException {
+        HashMap<String, GridCoverage2D> inputCoverages = new HashMap<>();
+        HashMap<String, GridCoverageReader> inputReaders = new HashMap<>();
+        GridCoverage2D dynamicAlphaSource = null;
+        int nonNullCoverages = 0;
+        for (int bIdx : selectedBandIndices) {
+            CoverageBand band = bands.get(bIdx);
+            List<InputCoverageBand> selectedBands = band.getInputCoverageBands();
+
+            // Peek for coverage name
+            String coverageName = selectedBands.get(0).getCoverageName();
+            if (!inputCoverages.containsKey(coverageName)) {
+                GridCoverage2DReader reader =
+                        SingleGridCoverage2DReader.wrap(delegate, coverageName);
+                inputReaders.put(coverageName, reader);
+                // Remove this when removing constraints
+                if (checker == null) {
+                    checker =
+                            new CoveragesConsistencyChecker(
+                                    reader, canSupportHeterogeneousCoverages);
+                } else {
+                    checker.checkConsistency(reader);
+                }
+                // bands selection parameter inside on final bands so they should not be propagated
+                // to the delegate reader
+                GeneralParameterValue[] filteredParameters = parameters;
+                if (parameters != null) {
+                    // creating a copy of parameters excluding the bands parameter
+                    filteredParameters =
+                            Arrays.stream(parameters)
+                                    .filter(
+                                            parameter ->
+                                                    !matches(parameter, AbstractGridFormat.BANDS))
+                                    .toArray(GeneralParameterValue[]::new);
+                }
+                GridCoverage2D coverage = reader.read(filteredParameters);
+                if (coverage == null) {
+                    if (handler.isHomogeneousCoverages()
+                            || handler.getEnvelopeCompositionType()
+                                    == EnvelopeCompositionType.INTERSECTION) {
+                        return null;
+                    }
+                } else {
+                    nonNullCoverages++;
+                }
+                if (dynamicAlphaSource == null && hasDynamicAlpha(coverage, reader)) {
+                    dynamicAlphaSource = coverage;
+                }
+                inputCoverages.put(coverageName, coverage);
+            }
+        }
+        InputAlphaNonNullCoverages inputAlphaNonNull =
+                new InputAlphaNonNullCoverages(
+                        bands, inputReaders, inputCoverages, dynamicAlphaSource, nonNullCoverages);
+        return inputAlphaNonNull;
+    }
+
+    public static class InputAlphaNonNullCoverages {
```

Comment:
Needs a better name... maybe ViewInputs? Also, does not seem like it needs to be public, package private should be enough.

Reply:
Changed and made package private.